MBS-11263 Provide logcat from Incomplete tests

dsvoronin opened the pull request:
Previously it was only available via LogcatBuffer, tailing logcat file during test execution; now it's needed to provide access with already populated string with logcat of Incomplete test

Logcat will be fully available only under "stdout" section of ReportViewer
Previous stdout/stderr separation was silly and misguiding (take a look and separation logic in diff)

Possible future improvements: to make logcat more readable some tags could be filtered out somewhere near LogcatAccessor logic, instead of feeding configuration with whitelist

dsvoronin committed:
MBS-11263 refactor logcat work to provide logcat from adb command

Previously it was only available via LogcatBuffer, tailing logcat file during test execution; now it's needed to provide access with already populated string with logcat of Incomplete test

add logcat command to Device

eugene-krivobokov commented on subprojects/common/report-viewer/src/main/kotlin/com/avito/report/internal/ReportsAddApiImpl.kt:
```diff
@@ -213,12 +209,12 @@ internal class ReportsAddApiImpl(private val client: JsonRpcClient) : ReportsAdd
-                "stderr" to stderr
+                "stdout" to logcat,
+                // non empty value causes warning sign to appear in UI
+                "stderr" to ""
```
How does it look?
Should we express it explicitly for a client instead of empty value?

dsvoronin replied:
![image](https://user-images.githubusercontent.com/1105133/120382679-9a526300-c32c-11eb-8d14-705bb59a0653.png)

Like this. We can but it will be an invalid link probably

dsvoronin replied:
I wanted to add "All logcat output goes to stdout" message to all stderr's, but it seems like a huge waste of traffic. It will be added to tens of thousands tests a day... 
Should be probably handled on report frontend side

eugene-krivobokov replied:
> Should be probably handled on report frontend side

Indeed

eugene-krivobokov commented on subprojects/test-runner/client/src/main/kotlin/com/avito/runner/scheduler/listener/ArtifactsTestListener.kt:
```diff
@@ -63,12 +63,20 @@ internal class ArtifactsTestListener(
-                TestResult.Incomplete(result)
+                TestResult.Incomplete(
+                    infraError = result,
+                    logcat = device.logcat(LOGCAT_LINES_TO_FETCH)
```
```suggestion
                    logcat = device.logcat(LOGCAT_LAST_LINES)
```
The current behavior is obvious. Let's explain the motivation.

eugene-krivobokov commented on subprojects/test-runner/client/src/main/kotlin/com/avito/runner/scheduler/listener/ArtifactsTestListener.kt:
```diff
@@ -63,12 +63,20 @@ internal class ArtifactsTestListener(
                     InfrastructureError.Unexpected(
                         IllegalStateException("Instrumentation executed Ignored test")
+                    ),
+                    logcat = Result.Failure(
```
It looks that Result is not suitable for this property.
Clients can't distinguish "empty\disabled" values from real errors.

dsvoronin replied:
Agreed, please take a look at new `LogcatResult`

eugene-krivobokov replied:
I see, `LogcatResult` looks solving the issue.
Can we use it here?

dsvoronin replied:
we should, but can't right now, it needs to be refactored in a way that most internal classes of logcat and report processing moved to `:client` module

sergioboi replied:
I don't get why we can't call `logcatBuffers.get(key)` when test is incomplete?

dsvoronin replied:
because logcat buffers needs start and stop events, and such incomplete tests doesn't have these callbacks called

eugene-krivobokov commented on subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/logcat/LogcatAccessor.kt:
```diff
@@ -0,0 +1,12 @@
+
+    /**
+     * Logcat will be populated with empty lists if not available;
+     * Check [isAvailable] to avoid useless work
```
It's easy to forget about "if" before usage. Nothing will remind or force to do it.
It'd be good to express this by types, e.g. sealed class.

dsvoronin replied:
https://github.com/avito-tech/avito-android/pull/1036#discussion_r643462103

eugene-krivobokov commented on subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/logcat/ResultLogcatAccessor.kt:
```diff
@@ -0,0 +1,13 @@
+
+    override fun getLogs(): Logcat {
+        return result.map { Logcat(it) }.getOrElse { Logcat.STUB }
+    }
```
Example for https://github.com/avito-tech/avito-android/pull/1036/files#r643426727

I will invoke `getLogs()` and get empty string.

dsvoronin replied:
https://github.com/avito-tech/avito-android/pull/1036#discussion_r643462103

dsvoronin committed:
MBS-9539 stdout and stderr merged into stdout field for ReportViewer

stderr default message points user to look for stdout

use Problem to describe logcat unavailability

sergioboi commented on subprojects/test-runner/client/src/main/kotlin/com/avito/runner/scheduler/listener/ArtifactsTestListener.kt:
```diff
@@ -84,3 +92,5 @@ internal class ArtifactsTestListener(
     }
 }
+
+private const val LOGCAT_LAST_LINES = 300
```
Why 300? Could we send all lines?

dsvoronin replied:
all lines count is unknown, and could be a big file; in fact:

```
adb logcat -g

main: ring buffer is 2 MiB (1 MiB consumed), max entry is 5120 B, max payload is 4068 B
system: ring buffer is 2 MiB (307 KiB consumed), max entry is 5120 B, max payload is 4068 B
crash: ring buffer is 2 MiB (0 B consumed), max entry is 5120 B, max payload is 4068 B
```
for our 29 emulator

we could alter this by `adb logcat -G <SIZE>` command

sergioboi commented on subprojects/test-runner/service/src/main/kotlin/com/avito/runner/service/worker/device/adb/AdbDevice.kt:
```diff
@@ -435,6 +435,28 @@ data class AdbDevice(
     )
 
+    override fun logcat(lines: Int): Result<String> {
+        return retryAction.retry(
```
One more `retry` action (

sergioboi commented on subprojects/test-runner/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/logcat/BufferLogcatAccessor.kt:
```diff
@@ -0,0 +1,16 @@
+
+import com.avito.android.Problem
+
+internal class BufferLogcatAccessor(private val logcatBuffer: LogcatBuffer?) : LogcatAccessor {
```
We could check `logcatBuffers.get(key)` and if there is `null` then call `device.logcat()`
